boot/spring_performance.go

```diff
@@ -0,0 +1,265 @@
+			trainingRunArgs = append(trainingRunArgs, "-Dspring.aot.enabled=true")
+		}
+
+		jarPath := s.AppPath
+
+		if s.ReZip {
+			jarDestDir := os.TempDir() + "/" + fmt.Sprint(time.Now().UnixMilli()) + "/jar-dest"
+			if err := os.MkdirAll(jarDestDir, 0755); err != nil {
+				return layer, fmt.Errorf("error creating temp directory for jar\n%w", err)
+			}
+			tempJarPath := filepath.Join(jarDestDir, "runner.jar")
+			if err := CreateJar(s.AppPath+"/", tempJarPath); err != nil {
+				return layer, fmt.Errorf("error recreating jar\n%w", err)
+			}
+			f, err := os.Open(tempJarPath)
+			if err != nil {
+				return layer, fmt.Errorf("error opening jar\n%w", err)
+			}
+			if err = sherpa.CopyFile(f, filepath.Join(layer.Path, "runner.jar")); err != nil {
+				return layer, fmt.Errorf("error copying jar\n%w", err)
+			}
+
+			jarPath = tempJarPath
+			os.RemoveAll(s.AppPath)
+		}
+
+		if err := s.springBootJarCDSLayoutExtract(jarPath); err != nil {
+			return layer, fmt.Errorf("error extracting Boot jar at %s\n%w", jarPath, err)
+		}
+		startClassValue, _ := s.Manifest.Get("Start-Class")
+
+		err := resetCreationTimeWithTouch(s)
+		if err != nil {
+			return libcnb.Layer{}, fmt.Errorf("error resetting file creation time for CDS\n%w", err)
+		}
+
+		trainingRunArgs = append(trainingRunArgs,
+			"-Dspring.context.exit=onRefresh",
+			"-XX:ArchiveClassesAtExit=application.jsa",
+			"-cp",
+		)
+		trainingRunArgs = append(trainingRunArgs, s.ClasspathString)
+		trainingRunArgs = append(trainingRunArgs, startClassValue)
+
+		var trainingRunEnvVariables []string
+		
+		javaToolOptions := sherpa.GetEnvWithDefault("CDS_TRAINING_JAVA_TOOL_OPTIONS", sherpa.GetEnvWithDefault("JAVA_TOOL_OPTIONS", ""))
+		if javaToolOptions != "" {
+			s.Logger.Bodyf("Training run will use this value as JAVA_TOOL_OPTIONS: %s", javaToolOptions)
+			trainingRunEnvVariables = append(trainingRunEnvVariables, fmt.Sprintf("JAVA_TOOL_OPTIONS=%s", javaToolOptions))
+		}
+
+		// perform the training run, application.dsa, the cache file, will be created
+		if err := s.Executor.Execute(effect.Execution{
+			Command: "java",
+			Env:     trainingRunEnvVariables,
+			Args:    trainingRunArgs,
+			Dir:     s.AppPath,
+			Stdout:  s.Logger.InfoWriter(),
+			Stderr:  s.Logger.InfoWriter(),
+		}); err != nil {
+			return libcnb.Layer{}, fmt.Errorf("error running build\n%w", err)
+		}
+
+		return layer, nil
+	})
+
+	if err != nil {
+		return libcnb.Layer{}, fmt.Errorf("unable to contribute spring-cds layer\n%w", err)
+	}
+	return layer, nil
+}
+
+func resetCreationTimeWithTouch(s SpringPerformance) error {
+	// we set the creation date to the buildpack default 1980/01/01 date; the date the layer will appear to be created at
+	return s.Executor.Execute(effect.Execution{
+		Command: "find",
+		Env:     []string{"TZ=UTC"},
+		Args:    []string{"./", "-exec", "touch", "-t", "198001010000.01", "{}", ";"},
+		Dir:     s.AppPath,
+		Stdout:  s.Logger.InfoWriter(),
+		Stderr:  s.Logger.InfoWriter(),
+	})
+}
```

Comment:
yeah, I started with the Go mtimes and unfortunately, I got slightly different results (CDS is super sensitive to timestamps) and since touch worked, and left if at that.

I can retry

Reply:
I worry about this from a performance perspective. The shelling out to find isn't bad, it's one command. but find itself is going to shell out to touch for every file in that directory. This is the app workspace so we have an unbounded number of files that could potentially be in here. If someone has a lot of little files, maybe you end up with 10,000. That's a lot of commands to run.

Even if you can't get `os.Chtimes` to work, which it seems like it should be possible, you might look at using `find` and piping to `xargs`. That can be more efficient because `xargs` can batch multiple files into one call to touch. So if you have 10k files, maybe it can do 500 files per call to touch (depends on the max arg size which I don't remember off hand), but that cuts the number of calls to 1 find, 1 xargs, and 20 touch.

If you didn't want to do the piping, you could walk in Go, batch up the files and then do something similar to xargs calling touch with multiple files.

Reply:
I added a commit that swaps this call out for `Chtimes` and it works!